### PR TITLE
JIT: optimize some cases of unused structs

### LINKED ISSUE #18710: Strange codegen with structure fields accesses
The C# repro code below shows non-uniformity in .NET Core 2.1 JIT behavior regarding (de)optimization of memory accesses having potential side effects. In short, test methods attempt to access all three fields of the structure instance of which only two first values are used.

```
[StructLayout(LayoutKind.Sequential)]
struct VT
{
    public static readonly int Size = Marshal.SizeOf<VT>();

    public int F1, F2, F3;

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public void Get(out int v1, out int v2, out int v3)
    {
        v1 = F1;
        v2 = F2;
        v3 = F3;
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void TestMethodInlining(VT* pVT)
{
    int v1, v2, v3;
    pVT->Get(out v1, out v2, out v3);
    Do(v1, v2);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void TestByRef(ref VT VTRef)
{
    int v1, v2, v3;
    v1 = VTRef.F1;
    v2 = VTRef.F2;
    v3 = VTRef.F3;
    Do(v1, v2);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void TestByPtr(VT* pVT)
{
    int v1, v2, v3;
    v1 = pVT->F1;
    v2 = pVT->F2;
    v3 = pVT->F3;
    Do(v1, v2);
}

static void Main(string[] args)
{
    byte* pDataBytes = stackalloc byte[VT.Size];
    VT* pVT = (VT*)pDataBytes;
    pVT->F1 = 1;
    pVT->F2 = 2;
    pVT->F3 = 3;

    TestMethodInlining(pVT);
    TestByRef(ref *pVT);
    TestByPtr(pVT);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void Do(int v1, int v2)
{
    Console.WriteLine("v1={0} v2={1}", v1, v2);
}
```

Disasm listings follow
------------------------------------------
**`TestMethodInlining()`**
Accesses all three fields. Access to 3rd field is still there even when `out v3` is replaced with `out _` . A side question - is it mandatory to adjust RSP anyhow for such methods?
```
--- Program2.cs ---
    pVT->Get(out v1, out v2, out v3);
000007FE78721590  sub         rsp,28h  
000007FE78721594  mov         edx,dword ptr [rcx]  
000007FE78721596  mov         eax,dword ptr [rcx+4]  
000007FE78721599  mov         ecx,dword ptr [rcx+8]  
    Do(v1, v2);
000007FE7872159C  mov         ecx,edx  
000007FE7872159E  mov         edx,eax  
000007FE787215A0  call        000007FE78721040  
000007FE787215A5  nop  
000007FE787215A6  add         rsp,28h  
000007FE787215AA  ret  
--- No source file -------------------------------------------------------------
```

------------------------------------------
**`TestByRef()`**
Access to the 3rd field is dropped. This looks strange to me since the side effects are supposedly the same as in previous case. Also, field access is generated if I wrap it with `Volatile.Read(ref VTRef.F3)`. Also, the same question regarding RSP as in previous case.
```
--- Program2.cs ---
    v1 = VTRef.F1;
000007FE78722AE0  sub         rsp,28h  
000007FE78722AE4  mov         edx,dword ptr [rcx]  
000007FE78722AE6  mov         eax,dword ptr [rcx+4]  
000007FE78722AE9  mov         ecx,edx  
000007FE78722AEB  mov         edx,eax  
000007FE78722AED  call        000007FE787215C0  
000007FE78722AF2  nop  
000007FE78722AF3  add         rsp,28h  
000007FE78722AF7  ret  
--- No source file -------------------------------------------------------------
```

------------------------------------------
**`TestByPtr()`**
Quite a mess! First two fields are accessed and copied to EDX + EAX like in case 2 (and unlike in case 1). And then it looks like entire structure is copied on stack for the reason quite obscure. [I believe I've seen other cases when the structure passed as method arg with **ref/in** modifiers is copied, but this could be separate issue]
```
--- Program2.cs ---
    v1 = pVT->F1;
000007FE78712B10  sub         rsp,38h  
000007FE78712B14  mov         edx,dword ptr [rcx]  
000007FE78712B16  mov         eax,dword ptr [rcx+4]  
000007FE78712B19  mov         r8d,dword ptr [rcx]  
000007FE78712B1C  mov         dword ptr [rsp+28h],r8d  
000007FE78712B21  mov         r8d,dword ptr [rcx+4]  
000007FE78712B25  mov         dword ptr [rsp+2Ch],r8d  
000007FE78712B2A  mov         ecx,dword ptr [rcx+8]  
000007FE78712B2D  mov         dword ptr [rsp+30h],ecx  
000007FE78712B31  mov         ecx,edx  
000007FE78712B33  mov         edx,eax  
000007FE78712B35  call        000007FE787115C0  
000007FE78712B3A  nop  
000007FE78712B3B  add         rsp,38h  
000007FE78712B3F  ret  
--- No source file -------------------------------------------------------------
```

-----------------------------------------

One observation is that JIT produces the same code for all three methods when I add `volatile` modifier to all three fields.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
In some cases CSC will use `ldobj; pop` to null check a pointer to
struct. This change avoids copying the struct value for such constructs.

Codegen may still redundantly null check, if there are multiple such checks
in a method.

Fixes #18710

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@briansull PTAL
cc @dotnet/jit-contrib 

No diffs in fx for crossgen or pmi. Possibly because CSC will sometimes use `ldfld` instead which is less costly and easier to dead code.

Diffs on example code from #18710 (added here as tests):
```
Summary:
(Lower is better)
Total bytes of diff: -216 (-25.50% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -194 : structpop2.dasm (-38.42% of base)
         -22 : structpop.dasm (-6.43% of base)
2 total files with size differences (2 improved, 0 regressed), 0 unchanged.
Top method improvements by size (bytes):
        -194 : structpop2.dasm - P:TestByPtr(long):int
         -22 : structpop.dasm - P:TestByPtr(long):int
2 total methods with size differences (2 improved, 0 regressed), 11 unchanged.
```
For instance:
```diff
 ; Assembly listing for method P:TestByPtr(long):int
 ; Emitting BLENDED_CODE for X64 CPU with AVX
 ; optimized code
 ; rsp based frame
 ; partially interruptible
 ; Final local variable assignments
 ;
-;  V00 arg0         [V00,T00] (  7,  7   )    long  ->  rcx
+;  V00 arg0         [V00,T00] (  5,  5   )    long  ->  rcx
 ;  V01 loc0         [V01,T02] (  2,  2   )     int  ->  rax
 ;  V02 OutArgs      [V02    ] (  1,  1   )  lclBlk (32) [rsp+0x00]
 ;  V03 tmp1         [V03,T01] (  2,  4   )     int  ->  rdx
-;  V04 tmp2         [V04    ] (  3,  6   )  struct (16) [rsp+0x28]   do-not-enreg[XS] addr-exposed
-;  V05 tmp3         [V05    ] (  1,  1   )     int  ->  [rsp+0x28]   do-not-enreg[X] addr-exposed V04.F1(offs=0x00) P-DEP
-;  V06 tmp4         [V06    ] (  1,  1   )     int  ->  [rsp+0x2C]   do-not-enreg[X] addr-exposed V04.F2(offs=0x04) P-DEP
-;  V07 tmp5         [V07    ] (  1,  1   )     int  ->  [rsp+0x30]   do-not-enreg[X] addr-exposed V04.F3(offs=0x08) P-DEP
 ;
-; Lcl frame size = 56
+; Lcl frame size = 40
 G_M47241_IG01:
-       sub      rsp, 56
+       sub      rsp, 40
 G_M47241_IG02:
        mov      edx, dword ptr [rcx]
        mov      eax, dword ptr [rcx+4]
-       mov      r8d, dword ptr [rcx]
-       mov      dword ptr [rsp+28H], r8d
-       mov      r8d, dword ptr [rcx+4]
-       mov      dword ptr [rsp+2CH], r8d
-       mov      ecx, dword ptr [rcx+8]
-       mov      dword ptr [rsp+30H], ecx
+       cmp      dword ptr [rcx], ecx
        mov      ecx, edx
        mov      edx, eax
        call     P:Do(int,int):int
        nop
 G_M47241_IG03:
-       add      rsp, 56
+       add      rsp, 40
        ret
-; Total bytes of code 48, prolog size 4 for method P:TestByPtr(long):int
+; Total bytes of code 26, prolog size 4 for method P:TestByPtr(long):int
```

File: src/jit/importer.cpp [hunk 1]
```diff
@@ -12549,7 +12551,20 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                             op1->AsCall()->gtCallType == CT_HELPER)
 #endif // UNIX_AMD64_ABI
                         {
-                            op1 = impGetStructAddr(op1, clsHnd, (unsigned)CHECK_SPILL_ALL, false);
+                            // If the value being produced comes from loading
+                            // via an underlying address, just null check the address.
+                            if (op1->OperIs(GT_FIELD, GT_IND, GT_OBJ))
```
Comment (mikedn) [inline src/jit/importer.cpp]: Hmm, any volatile concerns?

Comment (AndyAyersMS) [reply src/jit/importer.cpp]: Hmm, maybe so. Thanks for bringing that up.

A quick glance suggests the jit won't even record volatile prefixes for ldobj. Need to dig in some more.

Comment (mikedn) [reply src/jit/importer.cpp]: Ah, this case is for structs. It's not clear to me if volatile applies to structs. ECMA-335 doesn't seem to say that volatile is not allowed on structs but I don't think I've ever seen code using that.

But then it looks like the importer already drops popped volatile reads - #6172

**Comment by AndyAyersMS** [comment]
Looks like CI is still not running normally.

**Comment by briansull** [comment]
Do you know what is up with "Windows_NT x64 Checked CoreFX Tests"

I am also seeing one test in that leg report a failure as well.

**Comment by CarolEidt** [comment]
> Do you know what is up with "Windows_NT x64 Checked CoreFX Tests"

I opened #18840 for this yesterday, which has now been fixed with #18844.

**Comment by AndyAyersMS** [comment]
@briansull did you have any feedback?

**Pull request merged**